### ISSUE TITLE
Possible wrong implementation for the mutex

### ISSUE THREAD

**Issue by yuk1i** [asterinas/asterinas#1497]
When I first reviewed the mutex code, I was confused about why the sleep lock is implemented this way. 

Generally, the `unlock` routine should not be interruptable during two tasks: 1. finding the next possible waiter, and 2. pass the lock to it, or unlock. 

See Mutex design in Linux Kernel: https://docs.kernel.org/locking/mutex-design.html

However, the code shows that releasing the lock and finding the next possible process, are not an atomic operation. What I supposed is that the CPU is IRQ-on and task preemptable, after the `self.release_lock()` is called, which means a context-switch may happen there. I think this is a common mistake when designing a Mutex.

```
impl<T: ?Sized> Mutex<T> {
    /// Releases the mutex and wake up one thread which is blocked on this mutex.
    fn unlock(&self) {
        self.release_lock();
        self.queue.wake_one();
    }
}
```

Considering the following scenario:

P1 is calling `Mutex::unlock`, and it just releases the lock and is about to call `wake_one`, maybe a context switch happens here.
After P1 releases the lock, P2 gets the lock and immediately unlocks it. Then, there are two concurrent calls to `wake_one`.
At the same time, P3 and P4 have been added to the wait queue, before P2 gets and releases the lock. 

Please consider whether mutex works in this situation.

Also, I find another possible implementation flaw in the WaitQueue:

```
pub fn wait_until<F, R>(&self, mut cond: F) -> R
where
    F: FnMut() -> Option<R>,
{
    if let Some(res) = cond() {
        return res;
    }

    let (waiter, _) = Waiter::new_pair();
    let cond = || {
        self.enqueue(waiter.waker());
        cond()
    };
    waiter
        .wait_until_or_cancelled(cond, || Ok::<(), ()>(()))
        .unwrap()
}

pub fn wake_one(&self) -> bool {
    // Fast path
    if self.is_empty() {
        return false;
    }

    loop {
        let mut wakers = self.wakers.disable_irq().lock();
        let Some(waker) = wakers.pop_front() else {
            return false;
        };
        self.num_wakers.fetch_sub(1, Ordering::Release);
        // Avoid holding lock when calling `wake_up`
        drop(wakers);

        if waker.wake_up() {
            return true;
        }
    }
}
```

I'm uncertain about whether the closure of `cond` will re-enqueue the waker, even if it is waked up and re-evaluating the `cond()`. (I have not learned the Rust.)

**Comment by lrh2000** [comment]
> Generally, the `unlock` routine should not be interruptable during two tasks: 1. finding the next possible waiter, and 2. pass the lock to it, or unlock.

I'm not sure why this needs to be non-interruptible. AFAICK, for the Linux implementation, the two steps are also interruptible, see [the code](https://elixir.bootlin.com/linux/v6.11.5/source/kernel/locking/mutex.c#L906).

> P1 is calling `Mutex::unlock`, and it just releases the lock and is about to call `wake_one`, maybe a context switch happens here.
> After P1 releases the lock, P2 gets the lock and immediately unlocks it. Then, there are two concurrent calls to `wake_one`.
> At the same time, P3 and P4 have been added to the wait queue, before P2 gets and releases the lock.

One of P3 and P4 can get the mutex immediately, the other has to wait. This seems fine?

> I'm uncertain about whether the closure of `cond` will re-enqueue the waker, even if it is waked up and re-evaluating the `cond()`.

I do not understand this sentence. Could you please explain: (i) What is the expected behavior? (ii) What is the current/incorrect behavior?

**Comment by luvletter2333** [comment]
For thr first part:

```
raw_spin_lock(&lock->wait_lock);
	debug_mutex_unlock(lock);
	if (!list_empty(&lock->wait_list)) {
		/* get the first entry from the wait-list: */
		struct mutex_waiter *waiter =
			list_first_entry(&lock->wait_list,
					 struct mutex_waiter, list);

		next = waiter->task;

		debug_mutex_wake_waiter(lock, waiter);
		wake_q_add(&wake_q, next);
	}

	if (owner & MUTEX_FLAG_HANDOFF)
		__mutex_handoff(lock, next);

	raw_spin_unlock(&lock->wait_lock);

	wake_up_q(&wake_q);
```

Apparently, the unlock process in guarded by a spinlock(lock->wait_lock).

First, lock the spinlock, 
Second, if the wake list is not empty, get the next
Third, pass the lock to the next, or NULL
Fouth, unlock the spinlock. 
Fifth, if the lock is passed to next, wake up next

For the next two questions, I think that when a process is woken up, it will re-evaluation the cond(). However, even if cond() result is true, aka it gets the lock, the waker is still enqueued into the wait_list

**Comment by yuk1i** [comment]
For thr first part:

```
raw_spin_lock(&lock->wait_lock);
	debug_mutex_unlock(lock);
	if (!list_empty(&lock->wait_list)) {
		/* get the first entry from the wait-list: */
		struct mutex_waiter *waiter =
			list_first_entry(&lock->wait_list,
					 struct mutex_waiter, list);

		next = waiter->task;

		debug_mutex_wake_waiter(lock, waiter);
		wake_q_add(&wake_q, next);
	}

	if (owner & MUTEX_FLAG_HANDOFF)
		__mutex_handoff(lock, next);

	raw_spin_unlock(&lock->wait_lock);

	wake_up_q(&wake_q);
Apparently, the unlock process in guarded by a spinlock(lock->wait_lock).
```

First, lock the spinlock,
Second, if the wake list is not empty, get the next
Third, pass the lock to the next, or NULL
Fouth, unlock the spinlock.
Fifth, if the lock is passed to next, wake up next

Steps 1.4 are atomic since they are guarded by a spinlock.

For the next two questions, I think that when a process is woken up, it will re-evaluate the cond(). However, even if the result is true, aka it gets the lock, the waker is still enqueued into the wait_list. Then, it gets the lock and should leave the WaitQueue, but, it's waker is still in wait_list, right?

**Comment by lrh2000** [comment]
> Third, pass the lock to the next, or NULL
> [ .. ]
> Fifth, if the lock is passed to next, wake up next

This only happens if `MUTEX_FLAG_HANDOFF` is set. I believe the flag is optional, and the mutex should still work even if the flag is not set. I think this is a performance optimization rather than a correctness bug?

But yes, we currently do not have support for mutex handoff. Thanks for pointing this out.

> Then, it gets the lock and should leave the WaitQueue, but, it's waker is still in wait_list, right?

Yes, we currently do not dequeue the waker, so it stays in the queue. It is indeed a waste of resources.

**Comment by yuk1i** [comment]
what I want to say is not just about the waste of resources, I'm concerning whether another waking process will lose the wakeup signal

**Comment by lrh2000** [comment]
> what I want to say is not just about the waste of resources, I'm concerning whether another waking process will lose the wakeup signal

As far as I'm concerned, no. The `has_woken` flag will prevent this.

When a `Waiter` is dropped, it will call `Waker::close`:
https://github.com/asterinas/asterinas/blob/fe339fd81e9123205b0937bbcfd43c4b59451a03/ostd/src/sync/wait.rs#L275-L279
https://github.com/asterinas/asterinas/blob/fe339fd81e9123205b0937bbcfd43c4b59451a03/ostd/src/sync/wait.rs#L242-L246

After that we check the `cond` again:
https://github.com/asterinas/asterinas/blob/fe339fd81e9123205b0937bbcfd43c4b59451a03/ostd/src/sync/wait.rs#L221-L223

In `wake_one` we also skip dead wakers:
https://github.com/asterinas/asterinas/blob/fe339fd81e9123205b0937bbcfd43c4b59451a03/ostd/src/sync/wait.rs#L103-L105

"Not losing the wakeup event is one of the most important invariants of the `Waker`/`Wait`/`WaitQueue` APIs. If you have found that it is possible to lose wakeup events, please do report the concrete scenario. Thank you for auditing the related code!

**Comment by tangruize** [comment]
> what I want to say is not just about the waste of resources, I'm concerning whether another waking process will lose the wakeup signal

Lost wakeups won’t happen in this case. You can also refer to a similar discussion here: https://github.com/asterinas/asterinas/issues/1261

**Comment by yuk1i** [comment]
Thank you for your detailed replies.

However, I still manage to audit a possible race-condition in the wait.rs.

The `WaitQueue::is_empty` is not guarded by the `WaitQueue::wakers`. 

Consider Thread A is calling `enqueue`, it first locks the `wakers`, then `push_back`, then increase `num_wakers`. Thread B is calling `wake_one`, when it's checking during the fast path, maybe Thread A's waker is pushed back into the vector but `num_wakers` is not increased.

**Comment by lrh2000** [comment]
> However, I still manage to audit a possible race-condition in the wait.rs.

**Comment by lrh2000** [comment]
> However, I still manage to audit a possible race-condition in the wait.rs.

Thanks again for your auditing!



> Consider Thread A is calling `enqueue`, it first locks the `wakers`, then `push_back`, then increase `num_wakers`. Thread B is calling `wake_one`, when it's checking during the fast path, maybe Thread A's waker is pushed back into the vector but `num_wakers` is not increased.

 - Whenever thread A's waker is pushed, thread A will _later_ check if the condition is true.
 - Before thread B calls `wake_one`, it must make sure that it has _already_ set the condition.
 - So thread A _must_ be able to see the condition that thread B set earlier. This wakeup event is not lost.

Here is a more detailed explanation, taking into account that the CPU can reorder instructions. See the timeline below:
```
 [ Thread B ]                                [ Thread A ]

  B.1 set the condition
  B.2 call `wake_one`
  B.3 check `num_wakers`
      in `Release` memory order
      (=> `num_wakers == 0`)
                                             A.1 call `enqueue`
                                             A.2 increment `num_wakers`
                                                 in `Acquire` memory order
                                                 (`num_wakers == 0` => `num_wakers == 1`)
                                             A.3 check the condition
```

 - First of all, `B.3` and `A.2` are two atomic operations on the same variable (`num_wakers`), so [either `B.3` happens first or `A.2` happens first in the modification order](https://en.cppreference.com/w/cpp/atomic/memory_order#Modification_order).
 - If `A.2` happens first, then thread B will see the updated `num_wakers`. Everything will work normally.
 - If `B.3` happens first (which is the scenario you're describing, also the timeline shown above), `A.2` (as an atomic operation with the "acquire" memory order) will see the results produced by `B.3` (as an atomic operation with the "release" memory order); this implies that [thread A synchronizes with thread B](https://en.cppreference.com/w/cpp/atomic/memory_order#Synchronizes_with). As a result, when thread B later checks the condition, it must see the condition set by thread A, [guaranteed by the Release-Acquire ordering](https://en.cppreference.com/w/cpp/atomic/memory_order#Release-Acquire_ordering).

**Comment by yuk1i** [comment]
Yeah you are right. The `wait_until` semantics in Thread A still holds.

Sorry for the false report.

**Comment by tangruize** [comment]
> Whenever thread A's waker is pushed, thread A will later check if the condition is true.

In addition to @lrh2000's explanation, I want to clarify the exceptions where the condition is not checked after enqueuing when using `new_pair()`, `enqueue()` and `wait()` as standalone functions.

In this context, the condition is evaluated as false before the call to `enqueue()`. Because the condition can be set concurrently, it must be guarded by a lock. Therefore, during the execution of `enqueue()`, the lock is already acquired. If another process sets the condition to true and subsequently calls `wake_one()` or `wake_all()`, it must first acquire the lock.

There are two such standalone usages:

- [flock](https://github.com/asterinas/asterinas/blob/7ddfd42baa210656127044995d8707fde74fab4d/kernel/src/fs/utils/flock.rs#L120-L141)
- [range_lock](https://github.com/asterinas/asterinas/blob/7ddfd42baa210656127044995d8707fde74fab4d/kernel/src/fs/utils/range_lock/mod.rs#L232-L248)

**Comment by lrh2000** [comment]
> There are two such standalone usages:
>  * flock
>  * range_lock

@tangruize 

This may be a bit off topic..

Although I am not familiar with the code involved, I would like to say that `WaitQueue::enqueue` is `#[doc(hidden)]`, it is _not_ expected to be used as such. Could you please explain why you don't use normal APIs? This is something that violates the design points of the original APIs and should be fixed in my opinion.

**Comment by yuk1i** [comment]
Another possible flaw: There is a time window that: 1. A valid waker is enqueued, 2. The waiter is not dropped, aka `has_woken == false`.

Consider Thread A calls `wait_until`, enqueues a waiter, successfully win the `cond` and never calls `do_wait`. But, Thread A has not returned from `wait_until`, which the Waker in the queue is still valid.

Because there is no lock, IRQ-off or preempt-off during: 1. cond-checking and the Waiter dropping, 2. `self.has_woken` checking and `scheduler::unpark_target` in `Waker::wake_up`. (Consider an IRQ and a preemption happens exactly in these two places).

Thus, a waker (Thread B) may passes the checking `self.has_woken.swap(true) == false` in `Waker::wake_up`, and `scheduler::unpark_target` a waiter's taks (Thread A), which may be a running/sleeping/blocking task when the `unpark` is really called.

This _saved_ wakes up may cause spurious wake up, and please make sure such race spurious wake up has no side-effects on any subsystems.

**Comment by yuk1i** [comment]
Such case may be easy to verify in runtime, I think? 

Please consider add a panic-assert in `impl Drop for Waker`, to check that `has_woken == false`, which means, when the last reference to the Waker drops:

1. A wait - a wake-up did happens, `do_wait` reset this value to false.
2. The waker has never been waked up.

If the value is true, there must be a wake-up event wrongly delivered.

**Comment by lrh2000** [comment]
> spurious wake up

Spurious wakes are allowed and totally expected. I don't think they will cause anything wrong.

**Comment by lrh2000** [comment]
Thanks for your audit. If you still think our implementation is problematic, feel free to open a new issue, but I don't think we need to keep this one open forever.
